### ISSUE TITLE
Unable to view large images on OpenGL video output

### ISSUE THREAD

**Issue by rr-** [mpv-player/mpv#4264]
### mpv version and platform

```
mpv git-a993a871e (C) 2000-2017 mpv/MPlayer/mplayer2 projects
 built on Thu Mar 23 17:14:22 CET 2017
ffmpeg library versions:
   libavutil       55.34.101
   libavcodec      57.64.101
   libavformat     57.56.101
   libswscale      4.2.100
   libavfilter     6.65.100
   libswresample   2.3.100
ffmpeg version: 3.2.4
```

### Reproduction steps

Try to open large file:

```
wget -O ~/mars.jpg "https://upload.wikimedia.org/wikipedia/commons/2/28/'Calypso'_Panorama_of_Spirit's_View_from_'Troy'.jpg"
mpv ~/mars.jpg --no-config --loop-file=inf --profile=opengl-hq
```

### Expected behavior

I should see planet Mars

### Actual behavior

I see black window instead

I can see the image fine on `vo=x11`, however, `vo=opengl,x11` doesn't fall through onto `x11`.
Suspected cause - my GPU can't handle GL textures this big, and the error happens too late to be caught by the code that would fall back to `x11`.

### Log file

```
[playlist] Reading plaintext playlist.
Playing: /home/rr-/mars.jpg
 (+) Video --vid=1 (mjpeg)
[ffmpeg/video] mjpeg: overread 8
[ffmpeg/video] mjpeg: EOI missing, emulating
VO: [opengl] 23123x5093 yuv444p
[vo/opengl] after video texture creation: OpenGL error INVALID_VALUE.
[vo/opengl] after creating framebuffer texture: OpenGL error INVALID_VALUE
[vo/opengl] Error: framebuffer completeness check failed (error=36054).
[vo/opengl] after rendering: OpenGL error INVALID_FRAMEBUFFER_OPERATION.
[vo/opengl] after creating framebuffer texture: OpenGL error INVALID_VALUE
[vo/opengl] Error: framebuffer completeness check failed (error=36054).
[vo/opengl] after rendering: OpenGL error INVALID_FRAMEBUFFER_OPERATION.
[vo/opengl] after creating framebuffer texture: OpenGL error INVALID_VALUE
[vo/opengl] Error: framebuffer completeness check failed (error=36054).
[vo/opengl] after rendering: OpenGL error INVALID_FRAMEBUFFER_OPERATION.
[vo/opengl] after creating framebuffer texture: OpenGL error INVALID_VALUE
[vo/opengl] Error: framebuffer completeness check failed (error=36054).
[vo/opengl] after rendering: OpenGL error INVALID_FRAMEBUFFER_OPERATION.
[vo/opengl] after creating framebuffer texture: OpenGL error INVALID_VALUE
[vo/opengl] Error: framebuffer completeness check failed (error=36054).
[vo/opengl] after rendering: OpenGL error INVALID_FRAMEBUFFER_OPERATION.
V: 00:00:00 / 00:00:01 (0%)
[vo/opengl] after rendering: OpenGL error INVALID_FRAMEBUFFER_OPERATION.
[vo/opengl] after rendering: OpenGL error INVALID_FRAMEBUFFER_OPERATION.
[vo/opengl] after rendering: OpenGL error INVALID_FRAMEBUFFER_OPERATION.
[vo/opengl] after rendering: OpenGL error INVALID_FRAMEBUFFER_OPERATION.
[vo/opengl] after rendering: OpenGL error INVALID_FRAMEBUFFER_OPERATION.
[vo/opengl] after rendering: OpenGL error INVALID_FRAMEBUFFER_OPERATION.
[vo/opengl] after rendering: OpenGL error INVALID_FRAMEBUFFER_OPERATION.
[vo/opengl] after rendering: OpenGL error INVALID_FRAMEBUFFER_OPERATION.
[vo/opengl] after rendering: OpenGL error INVALID_FRAMEBUFFER_OPERATION.
[vo/opengl] after rendering: OpenGL error INVALID_FRAMEBUFFER_OPERATION.
V: 00:00:00 / 00:00:01 (0%)


Exiting... (Quit)
```

### Sample files

https://upload.wikimedia.org/wikipedia/commons/2/28/'Calypso'_Panorama_of_Spirit's_View_from_'Troy'.jpg

**Comment by wm4** [comment]
Such is life. I'm surprised FFmpeg still accepts this dimension - it must be close to the limit (which is about 10000*10000 pixels).

Best we could do is "pre-scaling" the image with libswscale.

**Comment by rr-** [comment]
Will that work with zooming and panning though?

**Comment by wm4** [comment]
You can try how that would work by using ``--vf=scale=...`` (vf_scale's options suck, there's no scale factor and you must pass a new size).

I think current hardware has texture limits from 4K x 4K to 32K x 32K (depends on the GPU).

**Comment by rr-** [comment]
Unfortunately it looks pretty bad, especially for highly detailed images...

It would be cool to be able to fit the texture on the viewport rather than source, and use sws to fit and translate the image into that viewport texture. (I think that's what already happens with `vo=x11`?)

Also if you zoom a large image close enough on `vo=x11`, the image gets stretched in a weird way. (I think the image size overflows?) The behavior seems to produce different artifacts for large image width and for large image height.

**Comment by wm4** [comment]
Well sure, with vo_x11 this will work better, because it involves cropping the image before libswscale, and then scaling the cropped image to screen size.

I don't know what's overflowing, or whether it's a mpv or libswscale bug.

**Comment by rr-** [comment]
Because `x11` seems to have no problem with images of such size, I'd use `x11` by default, but `x11` flickers a lot when switching between images, and `xv` (which seems to work similarly to `x11`) has suboptimal OSD scaling.

For this reason I think the optimal solution (good looks and presumably low effort) is to make `-vo=opengl,...` fall back to the next VO when `opengl` is unable to create the texture - that way, supplying `-vo=opengl,x11`, would trigger `x11` only for extreme images.

**Comment by haasn** [comment]
The “proper” fix would be to use tiled textures for images too large to fit inside a single texture, but unfortunately getting this to work properly would be pretty complicated.

**Comment by wm4** [comment]
Too complicated, except maybe as explicit downsample pass (which would disregard some correctness issues).

**Comment by qmega** [comment]
Out of curiosity, what would be the complications? I've been planning to write an image viewer and this issue had occurred to me. I figured the tiles would have to overlap by the radius of the scaler. What else would have to be done?

**Comment by wm4** [comment]
> I figured the tiles would have to overlap by the radius of the scaler. 

I guess this might work too. Still pretty complicated...

**Comment by haasn** [comment]
They would have to overlap, yes. But working that logic into vo_opengl would still be stupidly complicated, given that vo_opengl has more than enough things going on.

**Comment by ValZapod** [comment]
[vo/gpu/d3d11] Failed to create Texture2D

Also fails for d3d11. Ha

**Comment by BGodding** [comment]
Did you ever find a workaround for this? Running into the same issue when trying to use a playlist with mixed videos and images. Images > `GL_MAX_TEXTURE_SIZE` fail to load unless I am using `x11`, but the video performance on `x11` is poor. Currently I am using the command interface to dynamically change `vo` when a large image needs to be displayed, but that can cause some glitching during the transition.

**Comment by nickjj** [comment]
> Did you ever find a workaround for this? 

Yes sort of, I recently discovered mpv is an amazing image viewer but my GPU has limited memory (2 GB) and on Wayland just opening a few small images in parallel is enough to start crashing my compositor.

I added this to my config:

```ini
[extension.png]
vo=x11
[extension.jpg]
vo=x11
[extension.jpeg]
vo=x11
```

That takes care of 99% of the images I open, add more extensions as needed. Videos stay opengl accelerated.

It would be neat if mpv could introduce an option or abstracted extension that says "all images" or "all videos" but that's likely a separate discussion and line of thought.

**Comment by llyyr** [comment]
Since you're on wayland, you should use the wlshm VO. Also memory usage goes down a good amount after https://code.videolan.org/videolan/libplacebo/-/commit/6518435aa70150edad8d44c5e606363a35b23a53

**Comment by guidocella** [comment]
You can use a conditional profile like

```
[huge]
profile-cond=width > 16384 or height > 16384
vo=wlshm
```

> It would be neat if mpv could introduce an option or abstracted extension that says "all images" or "all videos" but that's likely a separate discussion and line of thought.

`profile-cond=p['current-tracks/video/image']`

**Comment by nickjj** [comment]
> Since you're on wayland, you should use the wlshm VO

Wow thanks, this is great.

By default mpv used ~300mb to open a 3500 x 2333 image that's 3.9mb on disk. That's verified by running `nvidia-smi` and seeing `mpv` listed as a process.

With `--vo=wlshm` mpv doesn't even show up which I guess means it's not using any GPU memory? The same result happens with `x11` too btw.

Likewise with a decent sized mp4, by default mpv would use around 40mb of GPU memory but with wlshm it's not listed. Additionally, wlshm does use ~5% more CPU on my machine for video playback but it's using less GPU as a trade off.

Is there any real harm in using wlshm for videos too?

**Comment by nickjj** [comment]
Sorry for the double message but it feels worthy. Using `wlshm` is deadly on Wayland because it shares memory with your compositor. I opened a few images and it made niri (Wayland compositor) balloon in GPU memory usage.

Closing mpv did not result in niri clearing the memory. I went out of GPU memory after opening 6 images and had to reboot.

I don't know if this is up to Wayland or the compositor or the application to address.

Mentioning it here because the only thing that works for me is using `vo=x11`, please don't remove this. It's marked as legacy but the other options do not work on Wayland with certain GPUs.

**Comment by llyyr** [comment]
wlshm doesn't have any hardware acceleration, you're seeing mpv use less resources but that's because the compositor is doing a lot of the work. It's also a low quality VO, so it shouldn't be used except if images can't be displayed on a GPU VO.

> Closing mpv did not result in niri clearing the memory. I went out of GPU memory after opening 6 images and had to reboot.

This seems like a Niri leak

**Comment by guidocella** [comment]
I can confirm there is no memory leak on wlroots.

**Comment by nickjj** [comment]
Thanks a lot for testing it, niri uses https://github.com/Smithay/smithay under the hood.

Edit: I tested my same system using KDE Plasma (Wayland) and it's not leaking either. Looks to be isolated to niri / smithay, or at least something related to those projects.

**Comment by mahkoh** [comment]
If you're on nvidia, that's a known nvidia bug. They have a softcoded list of compositor names to reduce the memory usage on those and niri might not be on that list.

It could also be a niri bug.

**Comment by mahkoh** [comment]
>Since you're on wayland, you should use the wlshm VO.

I was surprised that this worked at all since I've not seen a compositor that handles wl_shm other than by allocating the gpu buffer for you. So effectively the size limit continues to exist (16384 on all of my GPUs including llvmpipe but there are also older and newer GPU where the limit is smaller/larger) and wl_shm_pool.create_buffer will cause a protocol error if you exceed the limit.

But the image from the OP has a width of 21123. However, it turns out that mpv seems to limit its width to 16000 (which just so happens to be under the hardware limit) and scaling is done in the client instead of relying on wp_viewport and such.

Maybe it would be useful to change the limit of 16000 to something like xdg_toplevel.configure_bounds if the corresponding dimension in xdg_toplevel.configure is 0. That way, mpv with wlshm would not be different from any other wl_shm client running fullscreen (of which there are many, since in practice all QT applications use software rendering). This would allow wlshm to work with arbitrarily large images/videos and it would only use well-tested compositor code paths.

Edit: Now that I think about it, it's also possible that mpv is already using configure_bounds but this event is just not sent before the mpv window is mapped the first time, and so mpv falls back to 16000.

**Comment by nickjj** [comment]
> They have a softcoded list of compositor names to reduce the memory usage on those and niri might not be on that list.

Yep, I had already done that. Before doing that niri would use 1 GB on boot, now it uses 75 MB but it still leaks.

I ended up finding an open issue on smithay which has a known memory leak when closing windows. A substantial portion of memory that was used by the process never gets reclaimed. Cosmic which also uses smithay has its own issue with a ton of replies of folks having the same problem. Both NVIDIA and AMD are affected on a range of systems.

Funny enough mpv is being used to help reproduce it because it's so easy to open it with different types of `vo=` back-ends to demonstrate it's a problem when apps run in a number of different ways. I posted a bit near the bottom of https://github.com/Smithay/smithay/issues/1562.

**Comment by kasper93** [comment]
> However, it turns out that mpv seems to limit its width to 16000

16000 seems to be hardcoded maximum output window size in mpv. See: https://github.com/mpv-player/mpv/blob/c63fe8ab9556692cd921fbb627b42b7262edf9ee/video/out/win_state.c#L107-L108

The initial geometry is clamped to 16k.

> vo=wlshm

Generally I would suggest to avoid changing VO for big image support and instead pre-scale the input to manageable size with `--vf=scale`. You can use autoprofile.

Setting different VO is workaround and works by accident like @mahkoh noticed, because there is hardcoded clamp in window size in mpv.

**Comment by guidocella** [comment]
wlshm still display huge images increasing that clamp to 99000, so that only affects the window size. wlshm prescaling huge images also seems wrong since they look identical in `--vo=x11` and both are much sharper than pre-scaling with `--vo=gpu-next --vf="lavfi=[scale=w='if(gt(iw,128*128),128*128,iw)':h='if(gt(ih,128*128),128*128,ih)']"`.

EDIT: the correct filter that preserves the aspect ratio is `--vf="lavfi=[scale='min(16384,iw)':'min(16384,ih)':force_original_aspect_ratio=decrease]"`, still looks much blurrier than `--vo=x11/wlshm`.

**Comment by mahkoh** [comment]
It doesn't only affect the window size.

>wl_display#1: error 3: An error occurred while processing a request: Could not process a `wl_surface#8.commit` request: Could not apply a delayed commit: Could not create a shm texture: The image is too large

Other compositors handle this differently. For example, sway just doesn't render the texture and prints an error to stderr.

**Comment by guidocella** [comment]
It displays for me in sway after increasing that limit.

**Comment by mahkoh** [comment]
Because the window is sized down by the compositor after the first frame.

**Comment by kasper93** [comment]
> both are much sharper than pre-scaling with --vo=gpu-next --vf="lavfi=[scale=w='if(gt(iw,128*128),128*128,iw)':h='if(gt(ih,128*128),128*128,ih)']".

Just a note about this scaling. It doesn't preserve aspect ratio of the image, so bad quality is expected, by doing bad scaling.

**Comment by guidocella** [comment]
Yeah but I don't think anyone expects windows to be 16k+ pixels large? I mean when the window is scaled to the output and the image is unscaled and you pan around it is in the original dimensions as far as I can see. It doesn't look any worse than in pure image viewers.

**Comment by kasper93** [comment]
Clearly mpv is not suited to be used for large image viewing. It should be fixed, by autoscaling or otherwise, instead of suggesting workarounds that are sub-par and tickles compositors in wrong way.

**Comment by guidocella** [comment]
wlshm works perfectly. If smithay has a bug it doesn't mean it's subpar. To make it work out of the box gpu-next and gpu need to fail if the image dimensions are greater than the max GPU texture size, so that VO autoprobing will switch to software rendering automatically.

**Comment by kasper93** [comment]
> wlshm works perfectly.

It does not:
* `wl_display#1: error 3 ...` as @mahkoh reported
* It's not default VO and missing many features of default VO
* It needs tricks to trigger this VO for images
* It doesn't solve the problem, because by default mpv doesn't work for big images
* It works only on Wayland, not on macOS or Windows
* It indirectly forces software scaling, by using this VO, even though it should be done explicitly for any VO.

**Comment by nickjj** [comment]
> Clearly mpv is not suited to be used for large image viewing

Where would you draw the line on what a large image is btw? I've recently started using mpv as an image viewer and after using Windows for 25 years with IrfanView, mpv is the only tool I've found that gives me a comparable viewing experience (in a good way). Of course I use mpv for videos too.

Long story short, I often want to view images from my phone that are pretty big (roughly 4000 x 3000). I wouldn't necessarily want these scaled down in an automated way. I do use `autofit-larger=100%x100%` so all images always fit in my display though, I've opted into that behavior by adding it to my config.

**Comment by kasper93** [comment]
> Where would you draw the line on what a large image is btw?

Above you GPU texture size limit, which likely is 16384x16384, but it depends on your hardware. If you can open image with vo=gpu-next, you are golden and it's fine. This issue is about those huge images that happens to not fit GPU processing.

**Comment by nickjj** [comment]
Excellent, thanks for the clarification.

**Comment by guidocella** [comment]
I still don't get any wl_display error on sway. It seems that mahkoh is trying to make the windows itself larger than 16384 pixels? But most compositors probably limit the initial window size regardless, only Xorg lets you create bigger windows.

Like nickjj said if I make huge images unscaled I would rather view them without any loss of quality like in image viewers rather than scale automatically.

Xorg has software rendering through `--vo=x11` and from the manual Windows and macOS seem to have some options to force software rendering at least, but no dedicated VO so VO autoprobing wouldn't fix the issue there.
